Add greater than symbol to attribute escaping

johngodley opened the pull request:
Although `>` is a [valid attribute character](https://www.w3.org/TR/html5/syntax.html#elements-attributes), it is used by `wptexturize` in WordPress to split HTML tokens. Without escaping `wptexturize` will incorrectly tokenize a string, [causing problems](https://github.com/WordPress/gutenberg/issues/991). 

For example, add `test > thing` as the `alt` tag for an image in Gutenberg, and then view the post:

```html
<img src="kirby.jpg" alt="test > thing" class="wp-image-10"/>
```

Results in this HTML rendered in WordPress:

```html
<img src="kirby.jpg" alt="test > thing&#8221; class=&#8221;wp-image-10&#8243;/>
```

Encoding in Gutenberg prevents this problem while being transparent to the Gutenberg visual UI. It does have the effect that if you add a `>` into a block's attribute in HTML mode it will be replaced by `&gt;`, and this matches the same behaviour of `&` being replaced with `&amp;`.

The problem likely exists for other block types where an attribute is output directly based on user input. For example, custom class names.

Fixes #9915 
Fixes #8779

Note this won't fix any existing posts that have a `>` in an attribute. Re-saving these in Gutenberg should resolve the problem.

## How has this been tested?
Additional unit test added for the `>` symbol, and existing unit tests modified.

It can also be manually verified that adding `test > thing` to an image's `alt` tag is displayed as-is in the Gutenberg UI, and results in `test &gt; thing` appearing in the HTML.

```html
<img src="kirby.jpg" alt="test &gt; thing" class="wp-image-10"/>
```

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

johngodley committed:
Add greater than to attribute escaping

Although > is a valid attribute character it is used in wptexturize to split HTML tokens. Without escaping wptexturize will incorrectly tokenize a string, causing problems for everything else. Encoding it in Gutenberg prevents this problem while being transparent to the Gutenberg visual UI.

jorgefilipecosta reviewed (approved):
This PR fixed the problem in my tests, and now it looks like we are doing the same the classic editor does 👍 
I think we can merge this fix for now, but I wonder if this is not a bug that should be fixed in wptexturize.

aduth commented on packages/element/src/serialize.js:
```diff
@@ -295,13 +306,17 @@ export function escapeLessThan( value ) {
  * "[...] the text cannot contain an ambiguous ampersand [...] must not contain
  * any literal U+0022 QUOTATION MARK characters (")"
  *
+ * Additionally we escape the greater than symbol, as this is used by wptexturize to
```
As far as `@wordpress/element` is concerned, it should have no knowledge of a `wptexturize` existing. Further, there's nothing about this module which implies any relationship to being used in the context of a post (i.e. it can be used to produce HTML which would be injected in an admin screen). These modules are meant to be independently useful without consideration being strongly bound to any one specific context. Holding ourselves honest to this guideline will prevent biasing implementation toward our specific needs.

Now, with that said, it may still be our best option moving forward, though I'm also wondering a few things:

- Can this be fixed from `wptexturize`?
- Can this be fixed by running a string replace on the result generated by a serialize call? Or do we need to be granular about the replacement occurring within an attribute value?
- And if so, should we consider extension points to our handling of various parts of the serialization? An attribute value _visitor_, for example?

Considering WordPress' needs as foreign can help surface where these sorts of extension points might be broadly useful.

johngodley replied:
> these modules are meant to be independently useful

I wasn't aware of that, but it's good to know. This did seem like a surely-this-is-too-easy kind of fix and I was expecting it to raise a question or two!

My first thought was to look at `wptexturize`. It's a monster of an old function and seems complicated to fix without triggering consequences somewhere. Saying that, I'm unfamiliar with the process, and don't know if Gutenberg is going to be tied to a particular base WP version etc. A fix there is certainly ideal, but tackling it in Gutenberg seems more immediately practical.

I then looked at fixing it inside the blocks themselves, but this places the burden on the block creator.

I looked at the serialize code. I did consider doing a string replace, but fixing a regex bug with another regex seemed like double trouble! It should be possible though, and we don't need to be granular about it - any unencoded `>` in an attribute will break the post. This could leave the core Gutenberg code to produce clean HTML, and we then have a 'texturizing' layer.

I also looked further into the serialization code and wondered if we could do it at a higher level - somehow traverse the block output and add the extra encoding to attributes. It got complicated, and I then came across `escapeAttribute`, which as solutions go seemed ideally magical.

By extension points do you mean hooks? That would work, and we could target different environments. If `wptexturize` is fixed in a future WordPress, the `>` encoding could be removed etc

aduth replied:
>By extension points do you mean hooks? That would work, and we could target different environments. If wptexturize is fixed in a future WordPress, the > encoding could be removed etc

In my mind I had pictured something like an AST / other tree visitor, which could be passed as an option to serialize, or in the creation of a custom serializer (e.g. expose serialize factory method to create / override default behavior).

Prior art:

- https://github.com/ctrlplusb/react-tree-walker#visitor
- https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-visitors

nosolosw replied:
@WordPress/gutenberg-core I know there is a core ticket for this https://core.trac.wordpress.org/ticket/45387 It doesn't seem like it has progressed much. Shall we land this fix as a preventive measure while the core fix is ongoing?

mdawaffe replied:
> These modules are meant to be independently useful without consideration being strongly bound to any one specific context.

Using `>` in an attribute is valid HTML. Using `&gt;` in an attribute is also valid HTML. Important to this discussion: the two have the same meaning.

If we're aiming to be as general as possible, we should pick `&gt;` since it's useable in more contexts.

aduth replied:
> Using `>` in an attribute is valid HTML. Using `&gt;` in an attribute is also valid HTML. Important to this discussion: the two have the same meaning.

I don't disagree, but:

1. The expressed intention of this function was to apply the minimal transformations necessarily to transform an attribute value to be conformant as per the specification of what constitutes a [valid attribute value](https://w3c.github.io/html/syntax.html#elements-attributes).
2. The bug still remains in WordPress that [valid HTML becomes mangled upon save](https://core.trac.wordpress.org/ticket/45387) (whether from Gutenberg or elsewhere), but because there are few who feel comfortable touching the relevant texturize code, the impetus has been placed on integrators to implement workarounds, as we've done here.
3. None of this matters, as the pull request has been approved on the pragmatic consideration that this be solved.

mdawaffe replied:
You're right. Thank you for your patience and for humoring me :)

I wanted to use this package elsewhere and found I could not because of the details of its implementation. I attempting to understand why, I found this ticket. My comment was meant to be a springboard into that discussion elsewhere, but I instead derailed the conversation. My apologies.

aduth replied:
@mdawaffe Not at all derailed, certainly not. I think I'd stressed far too much the idea of being "generic" in my original comment, and it was a good prompt to reevaluate what I'd meant in terms of the stated purpose of a module / function.

johngodley commented:
Thanks for the nudge @nosolosw !

If the core ticket gets any movement then that may ultimately be the preferred route.

If not then based on Andrew's feedback it seems like a more comprehensive solution is preferred in the long run.

However, if this PR makes sense for the immediate future then I've rebased and fixed it up so it's working again. I can also create another ticket with the desired plans.

iseulde commented on packages/escape-html/src/index.js:
```diff
@@ -64,7 +75,7 @@ export function escapeLessThan( value ) {
  */
 export function escapeAttribute( value ) {
-	return escapeQuotationMark( escapeAmpersand( value ) );
+	return escapeGreaterThan( escapeQuotationMark( escapeAmpersand( value ) ) );
```
Could use a comment explaining why it diverges from HTML spec.

johngodley replied:
Ah yes, I think it got lost in the rebase. Added!

johngodley committed:
Add comment about encoding difference

aduth commented on packages/escape-html/src/index.js:
```diff
@@ -59,12 +70,15 @@ export function escapeLessThan( value ) {
  *
+ * Note we also escape the greater than symbol, as this is used by wptexturize to
+ * split HTML strings. This is a WordPress specific fix
+ *
```
I think it would also be good to include an actionable note for the future maintainer. Something like:

```suggestion
 *
 * Note that if a resolution for Trac#45387 comes to fruition, it is no longer
 * necessary for `escapeGreaterThan` to exist in this module.
 *
 * See: https://core.trac.wordpress.org/ticket/45387
```

aduth commented on packages/escape-html/src/index.js:
```diff
@@ -51,6 +51,17 @@ export function escapeLessThan( value ) {
+ *
+ * @return {string} Escaped string.
+ */
+export function escapeGreaterThan( value ) {
```
I think it would be best if we avoided exposing this on the public interface of the module, to avoid committing ourselves to future maintenance if it's arguably a workaround for a WordPress quirk.

Noting that it's exported for use by testing, alternatives could include one of either:

- Moving it to a separate file from which it is the default export, but is imported and then not subsequently exported by `index.js`
- Naming it as an [unstable API](https://github.com/WordPress/gutenberg/blob/83f03b18ee0cc29d3c309da2c5d9a39392eefa75/docs/contributors/coding-guidelines.md#experimental-apis) `__unstableEscapeGreaterThan`

aduth committed:
Improve escapeGreaterThan comment to refer to trac ticket

Co-Authored-By: johngodley <johng75@gmail.com>

johngodley committed:
Make escapeGreaterThan an unstable API function

We don’t want people using it

iseulde commented on packages/escape-html/src/index.js:
```diff
@@ -51,6 +51,17 @@ export function escapeLessThan( value ) {
+ *
+ * @return {string} Escaped string.
+ */
+export function __unstableEscapeGreaterThan( value ) {
```
Why are we even exposing it? If you're exposing it for tests, I'd recommend making a separate file and not export it in `index.js`.

aduth replied:
Related: https://github.com/WordPress/gutenberg/pull/9963#discussion_r237872160

mcsf commented:
Punting, as we seemingly haven't reached consensus yet.

johngodley committed:
Move escapeGreater into seperate file

It’s available internally and for testing, but isn’t exported outside of the package

johngodley committed:
Update version and changelog

gziolo commented:
> Punting, as we seemingly haven't reached consensus yet.

What are next steps? Should we land it or close it?

johngodley commented:
Ok, I’ve updated this so the unstable code is also in a separate file. I’ve left the tests in the same place as they seem to make more sense there. I’ve also included changes for the package version and history (pending a real date)

ellatrix commented on packages/escape-html/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 1.0.2 (2019-??-??)
```
Is this the right format? I saw previous additions with `(Unreleased)`, unsure if this is matched anywhere. Cc @gziolo.

gziolo replied:
`(Unreleased)` is the preferred one. At least it is documented this way:

https://github.com/WordPress/gutenberg/tree/master/packages#maintaining-changelogs

ellatrix commented:
Rebased this PR, but I have no idea why the tests are failing...

youknowriad committed:
Update the docs

westonruter commented:
Has the code here been reverted in any way? When adding `&gt;` or `&lt;` into an attribute value, the serializer forces it to be `>` and `<` respectively.

```html
<!-- wp:paragraph -->
<p><abbr title="1 > 2">False</abbr> and <abbr title="1 < 2">True</abbr></p>
<!-- /wp:paragraph -->
```

This gets rendered by WordPress as:

```html
<p><abbr title="1 > 2&#8243;>False</abbr> and <abbr title="1 < 2">True</abbr></p>
```

The “False and” text is erroneously consumed be the attribute:

![image](https://user-images.githubusercontent.com/134745/57719883-83515300-7635-11e9-8e54-362eda4a391f.png)

In this case, it seems to be a problem with `wptexturize()`, as it is converting the attribute closing `"`  into `&#8243;`.

But if the block serializer output `>` as `&gt;` then this symptom could be avoided.

aduth commented:
@westonruter Nothing more recent has occurred here to my knowledge. Could you create an issue which includes version information and sample code? I might wonder how your `abbr` elements are generated could have an impact on the serialization.

westonruter commented:
@aduth cheers. Added here: https://github.com/WordPress/gutenberg/issues/15636.
